File: plugins/collector/event/apps/bigdata/storm.go
package bigdata

import (
	"collector/cache/process"
	"collector/event/apps"
	"path/filepath"
	"regexp"
	"strings"
)

type Storm struct {
	version    string
	reg        *regexp.Regexp
	regVersion *regexp.Regexp
}

func (Storm) Name() string { return "storm" }

func (Storm) Type() string { return "bigdata" }

func (s Storm) Version() string { return s.version }

func (s *Storm) Match(p *process.Process) bool {
	if p.Name != "java" {
		return false
	}
	if s.reg == nil {
		s.reg = regexp.MustCompile(` org\.apache\.storm\.daemon\.`)
	}
	if res := s.reg.FindString(p.Argv); res == "" {
		return false
	}
	return true
}

func (s *Storm) Run(p *process.Process) (m map[string]string, err error) {
	s.version = ""
	if s.regVersion == nil {
		s.regVersion = regexp.MustCompile(`storm-client-(\d+\.)+(\d+)\.jar`)
	}

	m = make(map[string]string)
	// check run mode
	// or report the mainclass as well
	if strings.Contains(p.Argv, " org.apache.storm.daemon.nimbus.Nimbus") {
		m["run_mode"] = "nimbus"
	} else if strings.Contains(p.Argv, " org.apache.storm.daemon.ui.UIServer") {
		m["run_mode"] = "uiserver"
	} else if strings.Contains(p.Argv, " org.apache.storm.daemon.supervisor.Supervisor") {
		m["run_mode"] = "supervisor"
	} else {
		m["run_mode"] = "unknown"
	}

	// get the version
	var fds []string
	if fds, err = p.Fds(); err != nil {
		return
	}
	for _, fd := range fds {
		if filepath.Ext(fd) != ".jar" {
			continue
		}
		if res := s.regVersion.FindString(filepath.Base(fd)); res != "" {
			s.version = strings.TrimRight(strings.TrimPrefix(res, "storm-client-"), ".jar")
			break
		}
	}
	return
}

func init() { apps.Regist(&Storm{}) }

File: plugins/collector/src/event/apps/bigdata/storm.rs
use crate::event::apps::{find_version, jar_names_for_pid, AppProc, IApp};
use std::collections::HashMap;

pub struct Storm {
    version: String,
}

impl Storm {
    pub fn new() -> Self { Storm { version: String::new() } }
}

impl IApp for Storm {
    fn name(&self)     -> &'static str { "storm" }
    fn app_type(&self) -> &'static str { "bigdata" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool {
        p.name == "java" && p.argv.contains("org.apache.storm.daemon.")
    }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        self.version.clear();
        // "storm-client-2.4.0.jar"
        for jar in jar_names_for_pid(p.pid) {
            if jar.starts_with("storm-client-") && jar.ends_with(".jar") {
                let mid = jar.trim_start_matches("storm-client-").trim_end_matches(".jar");
                let ver = find_version(mid);
                if !ver.is_empty() { self.version = ver; break; }
            }
        }
        Ok(HashMap::new())
    }
}
